File: iosApp/iosApp/Presentation/Notification/NotificationViewModel.swift
//
//  NotificationViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/06.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class NotificationViewModel: ObservableObject {
    private let getNotificationFeedUseCase: GetNotificationFeedUseCase

    private let markNotificationReadUseCase: MarkNotificationReadUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    @Published private(set) var uiState = NotificationUiState.empty

    let event = PassthroughSubject<NotificationEvent, Never>()

    private func observeSession() {
        Task {
            do {
                for try await _ in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.loadNotifications()
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func loadNotifications() {
        uiState.isLoading = true
        uiState.errorMessage = nil

        Task {
            do {
                let result = try await getNotificationFeedUseCase.invoke(pageSize: 20)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? NotificationFeed {
                    uiState = NotificationUiState(
                        isLoading: false,
                        errorMessage: nil,
                        isLoggedIn: feed.isLoggedIn,
                        unreadCount: feed.unreadCount,
                        sections: feed.sections
                    )
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = "알림을 불러오지 못했습니다."
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = "알림을 불러오지 못했습니다."
            }
        }
    }

    private func handleNotificationTap(id: String, type: String, targetId: String?) {
        Task {
            do {
                let result = try await markNotificationReadUseCase.invoke(notificationId: id)

                if result is AppResultFailure {
                    uiState.errorMessage = "알림 상태를 업데이트하지 못했습니다."
                } else {
                    var updatedSections: [NotificationSection] = []
                    var unreadCount: Int32 = 0

                    for section in uiState.sections {
                        var items: [NotificationListItem] = []

                        for item in section.items {
                            if item.id == id {
                                let updatedItem = NotificationListItem(
                                    id: item.id,
                                    title: item.title,
                                    message: item.message,
                                    type: item.type,
                                    targetId: item.targetId,
                                    isRead: true,
                                    relativeTime: item.relativeTime
                                )
                                items.append(updatedItem)
                            } else {
                                items.append(item)
                            }
                        }
                        unreadCount += Int32(items.filter { !$0.isRead }.count)
                        updatedSections.append(
                            NotificationSection(
                                id: section.id,
                                title: section.title,
                                items: items
                            )
                        )
                    }
                    uiState.sections = updatedSections
                    uiState.unreadCount = unreadCount

                    if let targetId {
                        if type == "CAST_SHIFT" || type == "BIRTHDAY" {
                            event.send(.navigateToCast(id: targetId))
                        } else if type == "CAFE_NOTICE" || type == "CAFE_EVENT" || type == "CAFE_TABLE_COUNT_UPDATE" {
                            event.send(.navigateToCafe(id: targetId))
                        }
                    }
                }
            } catch {
                if Task.isCancelled { return }
                uiState.errorMessage = "알림 상태를 업데이트하지 못했습니다."
            }
        }
    }

    func onAction(_ action: NotificationAction) {
        switch action {
        case .backTapped:
            event.send(.navigateBack)
        case .notificationTapped(let id, let type, let targetId):
            handleNotificationTap(id: id, type: type, targetId: targetId)
        case .signInTapped:
            event.send(.navigateToSignIn)
        case .refresh:
            loadNotifications()
        }
    }

    init(
        getNotificationFeedUseCase: GetNotificationFeedUseCase = KoinInitializerKt.resolveGetNotificationFeedUseCase(),
        markNotificationReadUseCase: MarkNotificationReadUseCase = KoinInitializerKt.resolveMarkNotificationReadUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase()
    ) {
        self.getNotificationFeedUseCase = getNotificationFeedUseCase
        self.markNotificationReadUseCase = markNotificationReadUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase

        observeSession()
        loadNotifications()
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/notification/NotificationViewModel.kt
package com.hhp227.concafe.presentation.notification

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.usecase.GetNotificationFeedUseCase
import com.hhp227.concafe.domain.usecase.MarkNotificationReadUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase

class NotificationViewModel(
    private val getNotificationFeedUseCase: GetNotificationFeedUseCase,
    private val markNotificationReadUseCase: MarkNotificationReadUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(NotificationUiState.empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<NotificationEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private fun observeSession() {
        viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest {
                loadNotifications()
            }
        }
    }

    private fun loadNotifications() {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        viewModelScope.launch {
            when (val result = getNotificationFeedUseCase.invoke()) {
                is AppResult.Success -> {
                    _uiState.value = NotificationUiState(
                        isLoading = false,
                        errorMessage = null,
                        isLoggedIn = result.data.isLoggedIn,
                        unreadCount = result.data.unreadCount,
                        sections = result.data.sections
                    )
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = "알림을 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun handleNotificationClick(id: String, type: String, targetId: String?) {
        viewModelScope.launch {
            when (markNotificationReadUseCase.invoke(id)) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        val updatedSections = state.sections.map { section ->
                            section.copy(
                                items = section.items.map { item ->
                                    if (item.id == id) {
                                        item.copy(isRead = true)
                                    } else {
                                        item
                                    }
                                }
                            )
                        }
                        val unreadCount = updatedSections.sumOf { section ->
                            section.items.count { !it.isRead }
                        }

                        state.copy(
                            unreadCount = unreadCount,
                            sections = updatedSections
                        )
                    }
                    if (targetId != null) {
                        when (type) {
                            "CAST_SHIFT", "BIRTHDAY" -> _event.emit(NotificationEvent.NavigateToCast(targetId))
                            "CAFE_NOTICE", "CAFE_EVENT", "CAFE_TABLE_COUNT_UPDATE" -> _event.emit(NotificationEvent.NavigateToCafe(targetId))
                        }
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update { it.copy(errorMessage = "알림 상태를 업데이트하지 못했습니다.") }
                }
            }
        }
    }

    fun onAction(action: NotificationAction) {
        when (action) {
            NotificationAction.ClickBack -> {
                viewModelScope.launch {
                    _event.emit(NotificationEvent.NavigateBack)
                }
            }
            is NotificationAction.ClickNotification -> handleNotificationClick(action.id, action.type, action.targetId)
            NotificationAction.ClickSignIn -> {
                viewModelScope.launch {
                    _event.emit(NotificationEvent.NavigateToSignIn)
                }
            }
            NotificationAction.Refresh -> loadNotifications()
        }
    }

    init {
        observeSession()
        loadNotifications()
    }
}
